File: src/main/java/io/binac/leetcode/BinaryTreeZigzagLevelOrderTraversal.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).
 * <p>
 * <p>For example:
 * <p>Given binary tree [3,9,20,null,null,15,7],
 * <blockquote><pre>
 *         3
 *        / \
 *       9  20
 *         /  \
 *        15   7
 * </blockquote></pre>
 * return its zigzag level order traversal as:
 * <blockquote><pre>
 *     [
 *       [3],
 *       [20,9],
 *       [15,7]
 *     ]
 * </blockquote></pre>
 */
public class BinaryTreeZigzagLevelOrderTraversal {
    public static class Solution1 {
        private void preOrder(TreeNode root, int level, List<List<Integer>> result) {
            if (root == null) return;
            LinkedList<Integer> list;
            if (level == result.size())
                result.add(list = new LinkedList<>());
            else
                list = (LinkedList<Integer>) result.get(level);
            if (level % 2 == 0)
                list.addLast(root.val);
            else
                list.addFirst(root.val);
            preOrder(root.left, level + 1, result);
            preOrder(root.right, level + 1, result);
        }

        public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
            List<List<Integer>> result = new ArrayList<>();
            preOrder(root, 0, result);
            return result;
        }
    }
}

File: python3/leetcodepy/binary_tree_zigzag_level_order_traversal.py
"""
Given the root of a binary tree, return the zigzag level order traversal of its nodes' values. (i.e., from left to right, then right to left for the next level and alternate between).



Example 1:
https://assets.leetcode.com/uploads/2021/02/19/tree1.jpg

Input: root = [3,9,20,null,null,15,7]
Output: [[3],[20,9],[15,7]]

Example 2:

Input: root = [1]
Output: [[1]]

Example 3:

Input: root = []
Output: []


Constraints:

The number of nodes in the tree is in the range [0, 2000].
-100 <= Node.val <= 100
"""
from typing import List, Optional

from .utils import TreeNode


class Solution1:
    def pre_order(self, root: Optional[TreeNode], level: int, result: List[List[int]]):
        if root is None:
            return
        if len(result) == level:
            result.append([])
        result[level].append(root.val)
        self.pre_order(root.left, level + 1, result)
        self.pre_order(root.right, level + 1, result)

    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        self.pre_order(root, 0, result)
        for i in range(1, len(result), 2):
            result[i] = result[i][::-1]
        return result


class Solution2:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        if root is None:
            return result
        queue = [root]
        reverse = False
        while queue:
            level = []
            nex = []
            for node in queue:
                level.append(node.val)
                if node.left is not None:
                    nex.append(node.left)
                if node.right is not None:
                    nex.append(node.right)
            if reverse:
                level = level[::-1]
            result.append(level)
            reverse = not reverse
            queue = nex
        return result
